File: src/rht/raytracer/shapes/Sphere.java
package rht.raytracer.shapes;

import rht.raytracer.Ray;
import rht.raytracer.Vec3;

public class Sphere implements ShapeType {

    private final Vec3 centre;
    private final double radius;

    public Sphere(Vec3 centre, double radius) {
        this.centre = centre;
        this.radius = radius;
    }

    /**
     * Returns the distance at which the given ray intersects the object, or null if
     * it never does.
     */
    public Double intersect(Ray ray) {
        Vec3 v = ray.getOrigin().minus(centre);
        double dirDotV = ray.getDirection().dot(v);
        double discriminant = dirDotV * dirDotV - v.squared() + radius * radius;
        if (discriminant < 0) {
            return null;
        }
        double distance = -ray.getDirection().dot(v) - Math.sqrt(discriminant);
        if (distance >= 0) {
            return distance;
        } else {
            return null;
        }
    }

    /**
     * Returns the normal vector on the surface of a sphere at a given point.
     * 
     * @param pointOnSurface The point to find the normal at
     */
    public Vec3 normalAtPoint(Vec3 pointOnSurface) {
        return pointOnSurface.minus(centre).normalise();
    }
}

File: rustray/src/shapes/sphere.rs
use crate::{maths::vec::Vector, ray::Ray, shapes::Shape};

#[derive(Clone, Debug, PartialEq)]
pub struct Sphere {
    pub centre: Vector<3>,
    pub radius: f64,
}

impl Shape for Sphere {
    fn intersect(&self, ray: &Ray) -> Option<f64> {
        let v = ray.origin - self.centre;
        let dir_dot_v = ray.direction * v;
        let discriminant = dir_dot_v * dir_dot_v - v.squared() + self.radius * self.radius;
        if discriminant < 0.0 {
            return None;
        }
        let distance = -(ray.direction * v) - discriminant.sqrt();
        if distance >= 0.0 {
            Some(distance)
        } else {
            None
        }
    }

    fn normal_at(&self, point: Vector<3>) -> Vector<3> {
        (point - self.centre).normalise()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn intersect() {
        let sphere = Sphere {
            centre: Vector([0.0, 0.0, 0.0]),
            radius: 1.0,
        };

        let distance_perpendicular =
            sphere.intersect(&Ray::new(Vector([0.0, -5.0, 0.0]), Vector([0.0, 1.0, 0.0])));
        assert_eq!(distance_perpendicular, Some(4.0));

        let distance_parallel =
            sphere.intersect(&Ray::new(Vector([0.0, -5.0, 0.0]), Vector([1.0, 0.0, 0.0])));
        assert_eq!(distance_parallel, None);

        let distance_away = sphere.intersect(&Ray::new(
            Vector([0.0, -5.0, 0.0]),
            Vector([1.0, -1.0, -1.0]),
        ));
        assert_eq!(distance_away, None);

        let distance_inside = sphere.intersect(&Ray::new(
            Vector([0.0, 0.0, 0.0]),
            Vector([1.0, -1.0, -1.0]),
        ));
        assert_eq!(distance_inside, None);
    }

    #[test]
    fn normal_at_point() {
        let sphere = Sphere {
            centre: Vector([0.0, 0.0, 0.0]),
            radius: 2.0,
        };

        let normal = sphere.normal_at(Vector([0.0, 0.0, 2.0]));
        assert_eq!(normal, Vector([0.0, 0.0, 1.0]));
    }
}
